[back] Add current week stats in API stats endpoint

abhiram11 opened the pull request:
This PR is the beginning of the proposal to solve issue #974 and tries to work on both the Frontend and the Backend parts of the Requirement:

1) get this/last week comparison data from the backend . The `_days_delta` variables are temporarily changed due to testing reasons. I have added the `last_week` logics for `comparison` metric only, and not for `compared_entities`, if that's also required let me know.

2) show this data on the Frontend's `/comparison` page, with a progress bar that shows info regarding weekly goals and `loggedIn` user's personal contribution. The styling will be added later once the logic works !

Some screenshots for reference:

![comparison](https://user-images.githubusercontent.com/20269286/225590043-4d7c1940-3fbc-4355-aace-6749e2685851.JPG)
![stats](https://user-images.githubusercontent.com/20269286/225590051-a6449bf7-f921-4982-9767-78b3525d6461.JPG)

abhiram11 commented:
Can we add comparison count to redux storage, since we fetch it once in `HomeVideos` but I am calling the method `getPollStatsAsync` again in `Comparison.tsx` in `/pages`.

This is also done because I am not able to find any other way using which I can send `stats` props of `HomeVideos` to `Comparison.tsx`. I am not able to send these `props` through the `Button` that has `Start` as title on the Homepage.

lfaucon commented:
Hey! Thanks for the work. I think that the strategy you chose seems good.

> I have added the last_week logics for comparison metric only, and not for compared_entities, if that's also required let me know.

I would not say it is required, but I would be happy that we include it any way. That would make the code and data output more clean and uniform. Also we may use this data at some point for another feature.

> Can we add comparison count to redux storage, since we fetch it once in HomeVideos but I am calling the method getPollStatsAsync again in Comparison.tsx in /pages.

I agree that we could put the result of `getPollStatsAsync` only once in the Redux store. Do you agree @GresilleSiffle ?

> Should we add an object for the User in the Database itself to keep a check on the comparison that they have done? We would then need to define it in backend/core/models/user.py I believe.

I don't think that this much is needed. If you open the page https://tournesol.app/comparisons you can see that it shows the count of comparisons for hte user. This is obtained from the request to https://api.tournesol.app/users/me/comparisons/videos?limit=20&offset=0 which has a response `{"count":3422, "results":[ ... ]}`

lfaucon commented:
Hey! Thanks for the work. I think that the strategy you chose seems good.

> I have added the last_week logics for comparison metric only, and not for compared_entities, if that's also required let me know.

I would not say it is required, but I would be happy that we include it any way. That would make the code and data output more clean and uniform. Also we may use this data at some point for another feature.

> Can we add comparison count to redux storage, since we fetch it once in HomeVideos but I am calling the method getPollStatsAsync again in Comparison.tsx in /pages.

I agree that we could put the result of `getPollStatsAsync` only once in the Redux store. Do you agree @GresilleSiffle ?

> Should we add an object for the User in the Database itself to keep a check on the comparison that they have done? We would then need to define it in backend/core/models/user.py I believe.

I don't think that this much is needed. If you open the page https://tournesol.app/comparisons you can see that it shows the count of comparisons for hte user. This is obtained from the request to https://api.tournesol.app/users/me/comparisons/videos?limit=20&offset=0 which has a response `{"count":3422, "results":[ ... ]}`

lfaucon commented on backend/tournesol/views/stats.py:
```diff
@@ -104,13 +113,19 @@ def get(self, request):
                 datetime_lastedit__gte=time_ago(days=self._days_delta)
             ).count()
+            last_week_comparison_count = comparisons.filter(
+                datetime_lastedit__gte=time_ago(days=self._days_delta_week)
```
This would be a bit incorrect, because `timeago` used like this would return the datetime Monday at some hours:minutes depending on when we call it. I think we want instead to fix the datetime to the Monday at 00:00 of the current week

abhiram11 replied:
I will look into this, because since I was just passing the `number` of Days as the parameter for filter. I would need to create a more robust filter and also look into the `comparisons` array to see how to get more specific to get Monday 00:00

abhiram11 commented on backend/tournesol/serializers/stats.py:
```diff
@@ -15,6 +15,7 @@ class ComparedEntitiesStatisticsSerializer(Serializer):
 class ComparisonsStatisticsSerializer(Serializer):
     total = IntegerField()
     added_last_month = IntegerField()
+    added_last_week = IntegerField()
```
Will add this change!

abhiram11 commented:
> Hey! Thanks for the work. I think that the strategy you chose seems good.
> 
> > I have added the last_week logics for comparison metric only, and not for compared_entities, if that's also required let me know.
> 
> I would not say it is required, but I would be happy that we include it any way. That would make the code and data output more clean and uniform. Also we may use this data at some point for another feature.
> 
> > Can we add comparison count to redux storage, since we fetch it once in HomeVideos but I am calling the method getPollStatsAsync again in Comparison.tsx in /pages.
> 
> I agree that we could put the result of `getPollStatsAsync` only once in the Redux store. Do you agree @GresilleSiffle ?
> 
> > Should we add an object for the User in the Database itself to keep a check on the comparison that they have done? We would then need to define it in backend/core/models/user.py I believe.
> 
> I don't think that this much is needed. If you open the page https://tournesol.app/comparisons you can see that it shows the count of comparisons for hte user. This is obtained from the request to https://api.tournesol.app/users/me/comparisons/videos?limit=20&offset=0 which has a response `{"count":3422, "results":[ ... ]}`. I think we could make this API route enable to filter by date so that we get the desired count for the current week.

I'll make the code more uniform, won't take much of code lines I believe. Thanks for the last paragraph as well, I was not able to understand the related code base since I am not familiar with the whole backend flow just yet 😅

abhiram11 commented:
There are two variables that we can use for getting `current_week` comparison count for a user. They are `datetime_lastedit` and `datetime_add` in the `comparisons.py`. I think we should use `datetime_add` as the source of truth for our query for fetching the `current_week` count, would like to know your advice on this.

GresilleSiffle commented:
> I agree that we could put the result of getPollStatsAsync only once in the Redux store. Do you agree @GresilleSiffle ?

Yes, I agree. The top-most component currently responsible of fetching the statistics can dispatch an new action `stats/fetchStats` that will be handled by an async reducer (we can use the file `userSettingsSlice` as an example).

If it helps, here are the steps that should be followed to address everything we saw : )

**step 1**

- [ ] refactor the existing stats serializers to use `added_last_30_days` instead of `added_last_month`
- [ ] update the related tests, check by running the command `pytest`
- [ ] refactor the front end to use `added_last_30_days` instead of `added_last_month`
  - run `yarn update-schema` from the front end folder
  - run `./scripts/generate-services-from-openapi.sh` from the front end folder
  - replace the occurrences of  `added_last_month` by `added_last_30_days`
- [ ] do a manual test in the front end and commit

**step 2**

The process should be the same to add the new property `added_current_week` except that there is no occurrence to replace in the front end.

**step 3**

Use Redux instead of passing the statistics with props.

After these 3 steps the first pull request should be ready to be merge.

To create a component that displays the weekly collective goal and the personal contribution we will probably need to update the `users/me/comparisons` API to be able to filter comparisons by date (maybe a since GET parameter?, see the point **b** of technical specifications in https://github.com/tournesol-app/tournesol/issues/974). This enhancement and the related tests should be part of a second pull request, as the stats and comparisons APIs are not coupled.  

I propose to add the new component in a third PR firstly to limit the scope of this one to the back end refactor, and secondly to avoid too much round trips between the reviewers. I anticipate that the UX/UI of the new component could generate several feedbacks.

abhiram11 commented:
I missed out on running the `pytest` for this commit, I also did not follow up with the `yarn update-schema` and other steps in that mention as well, but I tried to add everything related to that manually to the code. This will help in refactoring the later commits by a ton!!

Please change the `ComparisonsStatistics.ts` manually since I am having as issue in committing it to the repo:
```
/* istanbul ignore file */
/* tslint:disable */
/* eslint-disable */

export type ComparisonsStatistics = {
    added_current_week: number;
    total: number;
    added_last_30_days: number;
};
```


I believe I did the `added_current_week` and refactor of `added_last_30_days` in this one, some proof-read will be required by myself as well. Here are some screenshots to visualize the current updates after I submitted 3 comparisons using test account:

The Frontend now receives the total of this month's comparison count as well!
![comparison](https://user-images.githubusercontent.com/20269286/226401430-98b90992-98ce-483f-a1ba-253994035544.JPG)

![stats](https://user-images.githubusercontent.com/20269286/226401211-60510127-0453-4b51-91e2-18376831944c.JPG)

GresilleSiffle commented:
> I tried pytest but it gave errors at places where I did not change the code. with errors such as no module named numpy, syntax errors, etc. I think there is issue in python version.

It looks like some Python packages are not installed. How did you install the back end? Are you using the Docker containers?

If you are using Docker, then the `pytest` command should be executed by the container:

```
docker exec -i tournesol-dev-api pytest
```

If you are using your own Python virtual environment, then you need to install the development requirements:

```bash
# from the back end directory (and while your virtualenv is activated)
pip install -r tests/requirements.txt
```

GresilleSiffle commented on frontend/src/features/comparisons/comparisonSlice.ts:
```diff
@@ -0,0 +1,36 @@
+
+export const initialState: PollStats = DEFAULT_POLL_STATS;
+
+export const comparisonSlice = createSlice({
```
This slice looks more like a `statsSlice` rather than a `comparisonSlice`.

I'd suggest to create a `statsSlice` instead, as the public statistics of the Tournesol API can contain different kind of data, like entities and users, and not only comparisons.

GresilleSiffle commented:
> The HomeVideos is the place from where the first fetch (I'll rename it to SET Comparison Data or something soon). I am thinking to apply a useEffect whenever there is a change in comparison data or if we SUBMIT a comparison on the My Comparison page. I would like your input on how should I move forward to solve this.

Just for my understanding, you are going to add the component "weekly collective goal" in this PR?

> I would like to know from which component(s) should the dispatch method for comparisonData be called for optimum performance.

You can dispatch the in the `HomeVideos` component, in the `useEffect` currently fetching the statistics.

> I can make a hook for this Comparison Slice just like useLoginState if it is required.

A hook can be useful if we want to refresh the statistics from different places in the application. You can use `useRefreshSettings` as example, available on the branch main.

abhiram11 commented:
@GresilleSiffle you're absolutely right, I am trying to add the  "weekly collective goal" (Frontend and Backend) and the Redux store implementation for this feature as well. Thank you for adding the references in the latest comment, I will look into them and will push a commit with the updates!

P.S. You seem to have missed some text in your comment "You can dispatch the in the HomeVideos component", would request you to add that to provide additional context.

abhiram11 commented:
This new commit has the few updates:
1) Refactored `comparisonSlice` to `statsSlice`
2) Added `useStatsRefresh` Hook for stats fetching, the demo is attached. The next step includes refactoring this hook at places where we have redundant code for fetching with the method name `getPollStatsAsync` used at -> `HomeVideos` and `usageStatsSection`.
3) The hook's `getPollStatsAsync` dispatch is also added to the `/comparison` "Compare 🌻" page where the Redux State of `statsSlice` is automatically updated whenever we click on the "Submit" button, as shown in the demo.

Here's a Demo for reference:

https://user-images.githubusercontent.com/20269286/227221052-6438e78d-4867-4bf1-ae9c-8e851e39d997.mp4

abhiram11 commented:
Glad to hear that, @GresilleSiffle ! I'll remove the comments and other irrelevant snippets.

Also, do you want me to remove the frontend code of progressBar and Buffer loaders? Or do you want me to style it instead? It should be noted that the user API comparison data etc is not being fetched!

abhiram11 commented:
As per the requirement, I have cleaned the comments from my code contribution for this PR. However, I have not removed the code for Progress Bar, Weekly Comparison Count Display, and related UI Code.

Would you like that code to be deleted or commented out? Thanks in advance!

lfaucon commented:
> Would you like that code to be deleted or commented out? Thanks in advance!

Thanks for the changes! A lot of it is ready to be merged and very well done 💯 

I prefer not to commit code that is commented out. The best would be if you can split the work that can be merged now into one PR, then the remaining changes should be added in one or multiple follow-up PRs/drafts. That would help us greatly to give feedback and move ahead.

abhiram11 commented:
That would be great, let me deletethe code and push the commit soon for a quick review. I have already replicated the frontend progress bare code for later commits and styling.

GresilleSiffle commented on backend/tournesol/tests/test_api_statistics.py:
```diff
@@ -105,10 +105,15 @@ def test_video_stats(self):
 
         comparison_count = video_poll["comparisons"]["total"]
-        last_month_comparison_count = video_poll["comparisons"]["added_last_month"]
+        #should this variable name be changed as well?
+        last_month_comparison_count = video_poll["comparisons"]["added_last_30_days"]
```
Yes let's do it.

It could be renamed simply `added_last_30_days` (or `last_30days_comparison_count`)

GresilleSiffle commented on backend/tournesol/views/comparison.py:
```diff
@@ -95,6 +95,7 @@ def post(self, request, *args, **kwargs):
         """
         Create a new comparison associated with the logged user, in a given
         poll.
+        Add the comparison count via this route/function, no need to apply this logic in updates sections etc.
```
Is this comment a kind of a personal reminder?

As this docstring is part of the public API documentation, I suggest to move this reminder elsewhere. If not it will be displayed to the API consumers, and I think this could be confusing.

See: https://api.tournesol.app/docs/#/users/users_me_comparisons_create

abhiram11 replied:
You're right, I was trying different approaches and steps to better my understanding of the backend flow, so I might have added some comments here and there and now am unaware of them since they were committed in super early commits. Apologies for that, they were just for my own reminder.

GresilleSiffle commented on backend/tournesol/views/stats.py:
```diff
@@ -4,13 +4,17 @@
 from dataclasses import dataclass
 from typing import List
 
+import logging
```
`logging` is imported but not used, let's not forget it : )

You can run `docker exec -i tournesol-dev-api ./scripts/ci/lint.sh` to easily spot little mistakes like this in the code.

abhiram11 replied:
Thank you for the advices as well apart from the work, they are really helpful!

GresilleSiffle commented on backend/tournesol/views/stats.py:
```diff
@@ -104,13 +118,26 @@ def get(self, request):
                 datetime_lastedit__gte=time_ago(days=self._days_delta)
             ).count()
+            current_week_comparison_count = comparisons.filter(
+                datetime_lastedit__gte=time_ago(days=daysParam, hours=hourParam, minutes=minuteParam, seconds=secondParam)
```
I think we can simplify this part:

```python
from datetime import datetime, time

# this is Monday 00:00:00 
last_monday = datetime.combine(time_ago(days=timezone.now().weekday()), time.min)

# then in the filter we use our datetime object directly
current_week_comparison_count = comparisons.filter(datetime_lastedit__gte=last_monday)
```

It will be less error-prone than having to manually work with days, hours, minutes and seconds.

abhiram11 commented:
> > There are two variables that we can use for getting current_week comparison count for a user. They are datetime_lastedit and datetime_add in the comparisons.py. I think we should use datetime_add as the source of truth for our query for fetching the current_week count, would like to know your advice on this.
> 
> You are right. We want the new comparisons made during the current week, so `datetime_add` is perfect.

This update seems to have been left... I am looking into it as well

abhiram11 commented on backend/tournesol/tests/test_api_statistics.py:
```diff
@@ -105,10 +105,13 @@ def test_video_stats(self):
         self.assertEqual(comparison_count, len(self._list_of_comparisons))
-        self.assertEqual(last_month_comparison_count, 2)
+        self.assertEqual(last_30days_comparison_count, 2)
+        self.assertEqual(current_week_comparison_count, 0)
```
What test case should I write here?

aidanjungo commented:
Hi, as we would like to keep the PR only backend related, could you remove all the code you change in the frontend and put this code you your PR #1480, I think it will also solve naturally the current conflicts

abhiram11 commented:
I have removed the frontend code, the backend code still works as required and can be reviewed by accessing `localhost/stats` after submitting a comparison.

lfaucon commented:
Thanks for the work @abhiram11 I'll take over fixing the last problem tomorrow and merge your contribution. I look forward to  this feature being complete 💯

lfaucon commented:
I didn't figure out how to push on the branch from your fork, so I opened the new PR: #1486

lfaucon commented:
> Hi, as we would like to keep the PR only backend related, could you remove all the code you change in the frontend and put this code you your PR https://github.com/tournesol-app/tournesol/pull/1480, I think it will also solve naturally the current conflicts

@aidanjungo That was a bit of a wrong advice because the backend changes were not backward compatible with the frontend (both should be changed at the same time, or the backend changes should be slightly different like I did in #1486 )

abhiram11 commented:
I have the old code as a backup as well, along with the commit history. Feel free to let me know how should it be added as the next contribution or so 👍🏻
